examples/tfidf.cpp

```diff
@@ -0,0 +1,156 @@
+  public:
+    using KeyT = std::string;
+    Document() = default;
+    explicit Document(const KeyT& t) : title(t){}
+    KeyT title;
+    std::vector<double> tf;
+    std::vector<double> tf_idf;
+    std::vector<std::string> words;
+    int total_words = 0;
+    std::string content;
+    const KeyT& id() const {
+        return title;
+    }
+
+    friend husky::BinStream& operator<<(husky::BinStream& stream, const Document& doc) {
+        stream << doc.title << doc.content << doc.tf << doc.tf_idf << doc.words << doc.total_words;
+        return stream;
+    }
+    friend husky::BinStream& operator>>(husky::BinStream& stream, Document& doc) {
+        stream >> doc.title >> doc.content >> doc.tf >> doc.tf_idf >> doc.words >> doc.total_words;
+        return stream;
+    } 
+};
+
+class Term {
+  public:
+    using KeyT = std::string;
+    Term() = default;
+    explicit Term(const KeyT& term) : termid(term) {}
+    KeyT termid;
+    double idf;
+    const KeyT& id() const {
+        return termid;
+    }
+};
+
+
+void load_mongodb(const std::string& server, const std::string& db,
+    const std::string collection, const std::string user, const std::string pwd) {
+    auto & document_list = husky::ObjListFactory::create_objlist<Document>();
+    auto parse = [&] (std::string& chunk) {
+        mongo::BSONObj o = mongo::fromjson(chunk);
+        Document doc(o.getStringField("title"));
+        doc.content = o.getStringField("content");
+        document_list.add_object(doc);
+    };
+    husky::io::MongoDBInputFormat inputformat;
+    inputformat.set_server(server);
+    inputformat.set_ns(db, collection);
+    inputformat.set_auth(user, pwd);
+    load(inputformat, parse);
+    auto& term_list = husky::ObjListFactory::create_objlist<Term>();
+    husky::lib::Aggregator<int> totalDoc;
+    auto& ac = husky::lib::AggregatorFactory::get_channel();
+    auto& ch = husky::ChannelFactory::create_push_combined_channel<int, husky::SumCombiner<int>>(document_list, term_list);
+    auto& ch2 = husky::ChannelFactory::create_push_channel<std::pair<std::string, int>>(document_list, term_list);
+    list_execute(document_list, {}, {&ch, &ch2, &ac}, [&](Document& doc) {
+        doc.words.resize(0);
+        std::vector<int> count;
+        if (doc.content.size()> 0) {
+            boost::char_separator<char> sep(" \t\n.,()\'\":;!?<>");
+            boost::tokenizer<boost::char_separator<char>> tok(doc.content, sep);
+            for (auto &w : tok) {
+                doc.words.push_back(w);
+                std::transform(doc.words.back().begin(), doc.words.back().end(), doc.words.back().begin(), ::tolower);
+            }
+            doc.total_words = doc.words.size();
+            std::sort(doc.words.begin(), doc.words.end());
+            int n = 0;
+            for (int i = 0, j = 0; i < doc.words.size(); i = j) {
+                for (j = i + 1; j < doc.words.size(); j++) {
+                    if (doc.words.at(i).compare(doc.words.at(j)) != 0)
```

Comment:
Do you mean this loop? It is O(n) since it breaks in the inner loop. The programs goes through the list only once.